### PR TITLE
Fix R/L bug in augmentation pipeline

### PR DESCRIPTION
**Pull request by julianrocha**
When we perform R/L augmentation, a right wrist becomes a left wrist but we do not account for this. Here is an example of the bug before these changes:
With the R/L flip (clearly wrong)
image
Without the R/L flip
image

This PR extends the aug pipeline to do custom R/L flip, and reshuffles the kp array accordingly.

### CODE REVIEW

**Comment by robertklee** [comment]
Have you noticed L/R labels being related to the actual person's left/right, or whether it's left/right from our view?

**Comment by julianrocha** [comment]
> Have you noticed L/R labels being related to the actual person's left/right, or whether it's left/right from our view?

It's the person's left/right. The ground truth image example with no flip in this PR demonstrates that.

File: data_augmentation.py [hunk 1]
```diff
@@ -223,6 +220,15 @@ def get_augmenter_pipeline(strength_enum):
 
     return aug_pipeline
 
+def flipRL(image, keypoints):
+    if random.random() < RL_FLIP:
+        return image, keypoints
+    flip = iaa.Fliplr(1.0)
+    flipped_img, flipped_kps = flip(image=image, keypoints=keypoints)
+    flat_map = lambda f, xs: [y for ys in xs for y in f(ys)] # https://dev.to/turbaszek/flat-map-in-python-3g98
```
Comment (robertklee) [inline data_augmentation.py:227]: Love this!!

File: data_generator_tests.py [hunk 2]
```diff
@@ -79,16 +85,16 @@ def display_img(annId):
...
-heatmap = y[:,:,1]
-print(X.max(), X.min(), X.mean())
-print(y.max(), y.min(), y.mean())
-hm = HeatMap(X,heatmap)
-hm.plot(transparency=0.5,show_axis=True,show_colorbar=True)
+for i in range(7,8):
+  print(COCO_KEYPOINT_LABEL_ARR[i])
+  heatmap = y[:,:,i]
+  hm = HeatMap(X,heatmap)
+  hm.plot(transparency=0.5,show_axis=True,show_colorbar=True)
```
Comment (robertklee) [inline data_generator_tests.py]: Small nit on indenting

**Review by robertklee — COMMENTED** [review]
I'm out right now but I will take a closer look in an hour or so. So far looks great! I love the lambda.

**Comment by robertklee** [comment]
> > Have you noticed L/R labels being related to the actual person's left/right, or whether it's left/right from our view?
> 
> 
> 
> It's the person's left/right. The ground truth image example with no flip in this PR demonstrates that.

That was what confused me, because it seems like the top image should be the person's left wrist, right?

Edit: never mind, it looks right now

File: data_augmentation.py [hunk 3]
```diff
@@ -223,6 +220,15 @@ def get_augmenter_pipeline(strength_enum):
...
+    flipped_fixed_kps = [flipped_kps[0]] + flat_map(lambda pair : (pair[1],pair[0]), zip(flipped_kps[1::2],flipped_kps[2::2]))
+    # Leave nose unchanged, flip every pair of eye, ear, etc...
+    return flipped_img, flipped_fixed_kps
```
Comment (nicolepeverley) [inline data_augmentation.py:230]: This looks great to me good catch! Just to confirm you're flipping it so that it remains in the keypoints order in constants?

**Review by nicolepeverley — APPROVED** [review]

File: data_augmentation.py (hunk 3 above)
Comment (julianrocha) [reply data_augmentation.py:230]: Yes, so:
`["nose","left_eye","right_eye","left_ear","right_ear","left_shoulder","right_shoulder","left_elbow","right_elbow","left_wrist","right_wrist","left_hip","right_hip","left_knee","right_knee","left_ankle","right_ankle"]`
becomes:
`["nose","right_eye","left_eye","right_ear","left_ear"...]`

This is necessary b/c when you mirror flip an image, a person's right side becomes their left side, so after fixing the coords, we need to swap labels with the other side.

Comment (nicolepeverley) [reply data_augmentation.py:230]: Ahh gotcha I wonder how this would work if we feed an augmented image into OKS or PCK if we can't directly compare to the ground truth.

File: data_augmentation.py [hunk 4]
```diff
@@ -11,6 +12,7 @@
 from imgaug.augmentables.kps import Keypoint, KeypointsOnImage
 
 from constants import ImageAugmentationStrength
+from constants import RL_FLIP
```
Comment (robertklee) [inline data_augmentation.py]: Nit,merge the two lines

File: data_augmentation.py (hunk 3 above)
Comment (robertklee) [reply data_augmentation.py:231]: We aren't applying augmentation to our validation or test set (and i'm not sure if we should). If we do, we can hold on to the augmented ground truth keypoints and use that as our ground truth. We will have to avoid running PCK on the training generator without accounting for this though

File: data_augmentation.py (hunk 4 above)
Comment (robertklee) [reply data_augmentation.py]: tip: you can right click in VSCode and there's actually a "sort imports" button that merges it for you

File: data_generator.py [hunk 5]
```diff
@@ -268,6 +268,9 @@ def __getitem__(self, idx):
                 # Perform data augmentation randomly
                 image_aug, kpsoi_aug = self.augmenter(image=transformed_img, keypoints=kpsoi)
 
+                # Perform a R/L augmentation randomly
```
Comment (robertklee) [inline data_generator.py]: This is a bit wordy, so feel free to reword.

```suggestion
                # Perform a R/L augmentation randomly, applying R/L flip to the labels as well to maintain the right order 
```

File: data_generator_tests.py [hunk 6]
```diff
@@ -56,19 +56,25 @@ def display_img(annId):
...
 import data_generator
 imp.reload(data_generator)
 from data_generator import DataGenerator
 
+
+representative_set_df = pd.read_pickle(os.path.join(DEFAULT_PICKLE_PATH, 'representative_set.pkl'))
 generator = DataGenerator(
-    df=val_df,
+    df=representative_set_df,
     base_dir=DEFAULT_TRAIN_IMG_PATH,
```
Comment (robertklee) [inline data_generator_tests.py]: Does `TRAIN_IMG_PATH` work here? (I think they're the same as `VAL_IMG_PATH`, right?)

File: data_generator_tests.py [hunk 7]
```diff
@@ -79,16 +85,16 @@ def display_img(annId):
...
-heatmap = y[:,:,1]
-print(X.max(), X.min(), X.mean())
-print(y.max(), y.min(), y.mean())
-hm = HeatMap(X,heatmap)
-hm.plot(transparency=0.5,show_axis=True,show_colorbar=True)
+for i in range(NUM_COCO_KEYPOINTS):
+  print(COCO_KEYPOINT_LABEL_ARR[i])
+  heatmap = y[:,:,i]
+  hm = HeatMap(X,heatmap)
+  hm.plot(transparency=0.5,show_axis=True,show_colorbar=True)
```
Comment (robertklee) [inline data_generator_tests.py:99]: small nit on indenting

**Review by robertklee — APPROVED** [review]
Looks good, and this is a super quick turnaround on this bug! 🚀

File: data_augmentation.py (hunk 3 above)
Comment (robertklee) [reply data_augmentation.py:230]: We aren't applying augmentation to our validation or test set (and i'm not sure if we should). If we do, we can hold on to the augmented ground truth keypoints and use that as our ground truth. We will have to avoid running PCK on the training generator without accounting for this though

File: data_augmentation.py (hunk 4 above)
Comment (julianrocha) [reply data_augmentation.py]: Didn't know this existed ty!

File: data_generator_tests.py (hunk 6 above)
Comment (julianrocha) [reply data_generator_tests.py]: They're the same but Imma change to  `VAL_IMG_PATH ` b/c it's more clear

**Comment by robertklee** [comment]
🚀 turnaround time! Looks like it's time to restart my training runs haha
